File: backend/src/modules/attendance/engine/tap-decision.ts
import { TapType } from '@prisma/client';

export interface OpenSessionInfo {
  id: string;
  loginAt: Date;
  siteId: string;
}

export interface LastTapInfo {
  clientEventTime: Date;
  tapType: TapType | null;
}

export type TapDecision =
  | { action: 'LOGIN' }
  | { action: 'LOGOUT'; sessionId: string }
  | { action: 'DUPLICATE'; cooldownRemainingSeconds: number }
  | {
      action: 'TOO_SOON';
      /** What the tap would have been had the gap already passed. */
      blocked: 'LOGIN' | 'LOGOUT';
      /** Seconds until the same scan would be accepted. */
      remainingSeconds: number;
      /** Minutes the worker has been in their current state. */
      elapsedMinutes: number;
    };

/**
 * Pure decision function: given the worker's current open session (if any),
 * their last tap, the site cooldown and the site safety gap, decide whether
 * this tap is a LOGIN, a LOGOUT, a DUPLICATE to ignore, or TOO_SOON to act on.
 *
 * Rules (docs/06-edge-cases.md #1, #4):
 *  - If the tap falls within the cooldown window of the last tap → DUPLICATE.
 *  - Else if the worker's state changed less than the safety gap ago → TOO_SOON.
 *  - Else if an open session exists → LOGOUT (closes it).
 *  - Else → LOGIN.
 *
 * The cooldown and the safety gap look alike but do different jobs. The
 * cooldown is measured from the last *tap* and is short — it swallows the badge
 * read that fires twice in the same second. The safety gap is measured from the
 * last *state change* and is long — it catches the read that lands a minute
 * later, once the cooldown has lapsed, which is how a queue of workers used to
 * get scanned back out shortly after arriving.
 *
 * Pass `safetyGapSeconds: 0` to switch the gap off — visitors are exempt,
 * because a ten-minute site visit is a normal visit.
 */
export function decideTap(
  tapTime: Date,
  cooldownSeconds: number,
  openSession: OpenSessionInfo | null,
  lastTap: LastTapInfo | null,
  safetyGapSeconds = 0,
): TapDecision {
  if (lastTap) {
    const elapsedMs = tapTime.getTime() - lastTap.clientEventTime.getTime();
    const cooldownMs = cooldownSeconds * 1000;
    if (elapsedMs >= 0 && elapsedMs < cooldownMs) {
      return {
        action: 'DUPLICATE',
        cooldownRemainingSeconds: Math.ceil((cooldownMs - elapsedMs) / 1000),
      };
    }
  }

  // The instant the worker entered their current state: the login that opened
  // the session they are in, or the logout tap that ended the last one. A tap
  // that predates it (a clock-skewed device, a late offline replay) is left to
  // the rules below rather than measured against a future event.
  const gapMs = Math.max(0, safetyGapSeconds) * 1000;
  const changedAt = openSession
    ? openSession.loginAt
    : lastTap?.tapType === 'LOGOUT'
      ? lastTap.clientEventTime
      : null;

  if (gapMs > 0 && changedAt) {
    const sinceChangeMs = tapTime.getTime() - changedAt.getTime();
    if (sinceChangeMs >= 0 && sinceChangeMs < gapMs) {
      return {
        action: 'TOO_SOON',
        blocked: openSession ? 'LOGOUT' : 'LOGIN',
        remainingSeconds: Math.ceil((gapMs - sinceChangeMs) / 1000),
        elapsedMinutes: Math.floor(sinceChangeMs / 60_000),
      };
    }
  }

  if (openSession) {
    return { action: 'LOGOUT', sessionId: openSession.id };
  }
  return { action: 'LOGIN' };
}

/** Haversine distance in metres between two lat/lng points. */
export function distanceMeters(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const R = 6371000;
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  return 2 * R * Math.asin(Math.min(1, Math.sqrt(a)));
}

/**
 * Decide whether photo verification should trigger for this tap.
 * `roll` is a 0-100 value (caller supplies randomness) so this stays pure.
 */
export function shouldVerifyPhoto(
  mode: 'ALWAYS' | 'NEVER' | 'RANDOM',
  randomPct: number,
  roll: number,
): boolean {
  if (mode === 'ALWAYS') return true;
  if (mode === 'NEVER') return false;
  return roll < randomPct;
}

File: mobile/lib/features/attendance/domain/tap_decision.dart
/// Pure offline tap-decision logic, mirroring the backend engine so the device
/// can decide LOGIN/LOGOUT/DUPLICATE/TOO_SOON instantly without a network
/// round-trip.
///
/// The last three are not produced by [decideTap] — the repository raises them:
/// `expired` when the worker's ID card has lapsed, so the tap is never queued;
/// `pendingApproval` when a punch was typed in by hand and now waits on a Safety
/// Officer; and `awaitingReview` when one is already waiting for that person.
enum TapAction { login, logout, duplicate, tooSoon, expired, pendingApproval, awaitingReview }

class OpenSession {
  const OpenSession({required this.id, required this.loginAt, required this.siteId});
  final String id;
  final DateTime loginAt;
  final String siteId;
}

class TapDecision {
  const TapDecision(
    this.action, {
    this.sessionId,
    this.cooldownRemainingSeconds = 0,
    this.blocked,
    this.remainingSeconds = 0,
    this.elapsedMinutes = 0,
  });

  final TapAction action;
  final String? sessionId;
  final int cooldownRemainingSeconds;

  /// For [TapAction.tooSoon]: what the scan would have been, and how long until
  /// it would be accepted.
  final TapAction? blocked;
  final int remainingSeconds;
  final int elapsedMinutes;
}

/// Decide whether a tap is a LOGIN, LOGOUT, a DUPLICATE to ignore, or TOO_SOON
/// to act on.
///
/// Rules (docs/06-edge-cases.md #1, #4):
///  - within cooldown of the last tap            -> DUPLICATE
///  - state changed less than the safety gap ago -> TOO_SOON
///  - else if an open session exists             -> LOGOUT
///  - else                                       -> LOGIN
///
/// The cooldown runs from the last *tap* and is short — it swallows a badge read
/// that fires twice in a second. The safety gap runs from the last *state
/// change* and is long — it catches the read that lands a minute later, once the
/// cooldown has lapsed. Pass `safetyGapSeconds: 0` to switch the gap off;
/// visitors are exempt and an override sets it to zero for that one scan.
TapDecision decideTap({
  required DateTime tapTime,
  required int cooldownSeconds,
  OpenSession? openSession,
  DateTime? lastTapTime,
  String? lastTapType,
  int safetyGapSeconds = 0,
}) {
  if (lastTapTime != null) {
    final elapsedMs = tapTime.difference(lastTapTime).inMilliseconds;
    final cooldownMs = cooldownSeconds * 1000;
    if (elapsedMs >= 0 && elapsedMs < cooldownMs) {
      final remaining = ((cooldownMs - elapsedMs) / 1000).ceil();
      return TapDecision(TapAction.duplicate, cooldownRemainingSeconds: remaining);
    }
  }

  // When the worker entered the state they are in now: the login that opened
  // their session, or the logout that ended the last one.
  final gapMs = (safetyGapSeconds < 0 ? 0 : safetyGapSeconds) * 1000;
  final changedAt = openSession != null
      ? openSession.loginAt
      : (lastTapType == 'LOGOUT' ? lastTapTime : null);

  if (gapMs > 0 && changedAt != null) {
    final sinceChangeMs = tapTime.difference(changedAt).inMilliseconds;
    if (sinceChangeMs >= 0 && sinceChangeMs < gapMs) {
      return TapDecision(
        TapAction.tooSoon,
        sessionId: openSession?.id,
        blocked: openSession != null ? TapAction.logout : TapAction.login,
        remainingSeconds: ((gapMs - sinceChangeMs) / 1000).ceil(),
        elapsedMinutes: sinceChangeMs ~/ 60000,
      );
    }
  }

  if (openSession != null) {
    return TapDecision(TapAction.logout, sessionId: openSession.id);
  }
  return const TapDecision(TapAction.login);
}
